core/src/main/java/org/elasticsearch/search/aggregations/AggregatorFactories.java

```diff
@@ -185,14 +188,50 @@ private void resolvePipelineAggregatorOrder(Set<String> aggFactoryNames, Map<Str
                 for (String bucketsPath : bucketsPaths) {
-                    List<String> bucketsPathElements = AggregationPath.parse(bucketsPath).getPathElementsAsStringList();
-                    String firstAggName = bucketsPathElements.get(0);
-                    if (bucketsPath.equals("_count") || bucketsPath.equals("_key") || aggFactoryNames.contains(firstAggName)) {
+                    List<AggregationPath.PathElement> bucketsPathElements = AggregationPath.parse(bucketsPath).getPathElements();
+                    String firstAggName = bucketsPathElements.get(0).name;
+                    if (bucketsPath.equals("_count") || bucketsPath.equals("_key")) {
+                        continue;
+                    } else if (aggFactoriesMap.containsKey(firstAggName)) {
+                        AggregatorFactory aggFactory = aggFactoriesMap.get(firstAggName);
+                        for (int i = 1; i < bucketsPathElements.size(); i++) {
+                            PathElement pathElement = bucketsPathElements.get(i);
+                            String aggName = pathElement.name;
+                            if ((i == bucketsPathElements.size() - 1) && (aggName.equalsIgnoreCase("_key") || aggName.equals("_count"))) {
+                                break;
+                            } else {
+                                // Check the non-pipeline sub-aggregator
+                                // factories
+                                AggregatorFactory[] subFactories = aggFactory.factories.factories;
+                                boolean foundSubFactory = false;
+                                for (AggregatorFactory subFactory : subFactories) {
+                                    if (aggName.equals(subFactory.name)) {
+                                        aggFactory = subFactory;
+                                        foundSubFactory = true;
+                                        break;
+                                    }
+                                }
+                                // Check the pipeline sub-aggregator factories
+                                if (!foundSubFactory && (i == bucketsPathElements.size() - 1)) {
+                                    List<PipelineAggregatorFactory> subPipelineFactories = aggFactory.factories.pipelineAggregatorFactories;
+                                    for (PipelineAggregatorFactory subFactory : subPipelineFactories) {
+                                        if (aggName.equals(subFactory.name())) {
+                                            foundSubFactory = true;
+                                            break;
+                                        }
+                                    }
+                                }
+                                if (!foundSubFactory) {
+                                    throw new IllegalStateException("No aggregation [" + aggName + "] found for path [" + bucketsPath + "]");
```

Comment:
I think this should rather be an IllegalArgumentException? (ie. the path argument is illegal?)